Adds htaccess file viewer to tools

jocastaneda opened the pull request:
As the title says, it adds the `.htaccess` file viewer to the tools tab if the file exists and has any custom or additional text. It also provides the ability to copy the text just in case a user needs to copy/paste that.

fixes: https://github.com/WordPress/healtch-check/issues/178

jocastaneda committed:
Adds htaccess file viewer to tools

With this commit, it adds the ability to view the .htaccess file the site
has if it exists and has additional rules added. This does not take into
account any comments the file might have since it checks only for core
added rules in the file.

fixes: https://github.com/WordPress/healtch-check/issues/178

jocastaneda committed:
Convert to unix line endings

Clorith commented on src/includes/class-health-check-htaccess-viewer.php:
```diff
@@ -0,0 +1,55 @@
+		}
+
+		// It does so let's get the contents
+		$htaccess_file = file_get_contents( ABSPATH . '/.htaccess' );
```
I would add a check that we can actually read the file contents, as in some scenarios, it is "locked" from the web user.

jocastaneda replied:
true, will look into this either today or tomorrow.

Clorith commented on src/includes/class-health-check-htaccess-viewer.php:
```diff
@@ -0,0 +1,55 @@
+
+		// Return since it is only core added rules.
+		if ( '' === $filtered_htaccess_content ) {
+			return $tabs;
```
I'm wondering if we should allow the displaying of the content, even if it's only got core content, avoiding conditional displaying as often as possible for a clearer user experience?

jocastaneda replied:
So, display it regardless?

I'm fine with that. We'd need to add a message stating something for it too.

Clorith reviewed (commented):
I like it, just a few discussion points I've marked that I'd like to hear your thoughts on before moving on here.

jocastaneda committed:
Merge latest master to branch

jocastaneda committed:
Adds check for read permission

This adds a `is_readable()` check for the .htaccess file. If it is
readable it will continue and output the content.

jocastaneda commented:
@Clorith I think when there is an `.htaccess` file we will need to add at least something stating hey, there are core added and if there are more, let the user know as well.

Clorith commented:
A short explanation of what the htaccess file does will probably suffice I'm thinking, since there's a "WordPress" wrapper around the core rules them selves, I wouldn't delve too deeply into semantics, as that falls a bit out of scope in my opinion.

jocastaneda committed:
Merge latest master to branch 178-htaccess-viewer

Clorith committed:
Change plugin link to https

Clorith committed:
Add REST API checks to the site status

Clorith committed:
PHP Extension tests

Clorith committed:
Add Plugin/Theme update checks to the site status tests

Clorith committed:
Check for HTTP requests being blocked

Clorith committed:
Add Twenty Nineteen as a default theme

Clorith committed:
Fix indentations for the debug text copy area

Clorith committed:
Various improvements based on feedback

Clorith committed:
Test if the REST API is able to access query parameters

Clorith committed:
Add docker for easier testing on the go

Clorith committed:
Don't test how long `dotorg_communication` takes

Clorith committed:
Don't include conditional messages in email tests

Clorith committed:
Fix tab indentations for stylelint

Clorith committed:
FRix docker localhost certs

Clorith committed:
Version bumped to match svn

Clorith committed:
Remove now unused admin header call

Clorith committed:
PHPCS fixes after cherrypicking.

Clorith committed:
Add nonces and capability checks to ajax actions

Clorith committed:
Add capability checks to the MU plugin

Clorith committed:
Send nonce with the `seen warning` event.

Clorith committed:
Remove capability check running too early

Clorith committed:
Add IP as salt to troubleshooting sessions

Clorith committed:
Make sure disabling a plugin doesn't cause a WSOD on error

Clorith committed:
Address PHPCS warnings

Clorith committed:
Update readme with new changelog

Clorith committed:
Version bump

Clorith committed:
Make sure tests have REMOTE_ADDR server variables available.

Clorith committed:
Use the new IP salt for MU plugin tests as well.

Clorith committed:
Merge branch 'master' into develop

Clorith committed:
Improve contributor documentation for the new `develop` branch and include a PR template so we get a more unified PR experience.

Clorith committed:
Fix readme typo and wrong markdown order on links

cjhaas committed:
Fix for plugin_data missing a slug (#235)

Fixes #234

Clorith committed:
Add missing REST API nonce header for REST tests (#238)

Clorith committed:
Fix backup warning not being dissmissible

Fixes #240

Soean committed:
Fix typos (#230)

Clorith commented:
Would you please change the branch to `develop` before this can be implemented, to suit the plugins new development flow.

siliconforks committed:
Make sure all files are not directly accessible (#241)

Resolves #237.

jocastaneda commented:
Changed it. Will finally get a chance to work on this tonight! 😍

jocastaneda committed:
Merge branch 'develop' into 178-htaccess-viewer

Pulls in latest

iamshouvikmitra committed:
Fixed misspelled words in comments (#224)

jocastaneda committed:
Merge remote upstream develop branch to local

jocastaneda committed:
Addresses feedback

* uses WP_Filesystem()
* uses `esc_html()`
* changes phrasing and removes the usage of `here`

jocastaneda commented on src/includes/class-health-check-debug-data.php:
```diff
@@ -461,9 +461,16 @@ static function debug_data( $locale = null ) {
 				'label' => __( 'htaccess rules', 'health-check' ),
-				'value' => ( ! empty( $filtered_htaccess_content ) ? __( 'Custom rules have been added to your htaccess file', 'health-check' ) : __( 'Your htaccess file only contains core WordPress features', 'health-check' ) ),
+				/* translators: admin url */
+				'value' => ! empty( $filtered_htaccess_content ) ? sprintf( __( 'Custom rules have been added to your htaccess file, you can view the contents <a href="%s">here</a>', 'health-check' ), esc_url( $tools_tab ) ) : __( 'Your htaccess file only contains core WordPress features', 'health-check' ),
```
Addressed in lates commit

jocastaneda commented on src/includes/class-health-check-htaccess-viewer.php:
```diff
@@ -0,0 +1,53 @@
+			);
+		} else {
+			// it is a file.
+			$htaccess_file = file_get_contents( ABSPATH . '/.htaccess' );
```
Addressed in latest commit

jocastaneda commented on src/includes/class-health-check-htaccess-viewer.php:
```diff
@@ -0,0 +1,53 @@
+			$htaccess_file = file_get_contents( ABSPATH . '/.htaccess' );
+
+			// Remove WordPress rules.
+			$filtered_htaccess_content = trim( preg_replace( '/\# BEGIN WordPress[\s\S]+?# END WordPress/si', '', $htaccess_file ) );
```
Yes. Can't recall why it was there to begin. It has been removed in the latest commit as well

jocastaneda commented on src/includes/class-health-check-htaccess-viewer.php:
```diff
@@ -0,0 +1,53 @@
+				<p>
+					<button type="button" class="button button-primary health-check-copy-field"><?php esc_html_e( 'Mark field for copying', 'health-check' ); ?></button>
+				</p>
+				<textarea id="htaccess-contents" class="widefat" rows="15"><?php echo htmlspecialchars( file_get_contents( ABSPATH . '.htaccess' ) ); ?></textarea>
```
In the latest it uses `esc_html`. I'm not super picky about it

carl-alberto committed:
Added prefix wp- to cookie name for added hosting compatibility (#253)

Clorith committed:
UI/UX redesign, first iteration, needs work on some action items still

Clorith committed:
New UI/UX continued work, also introduces a number of a11y improvements and additions to Gutenberg-esqueue styles.

Clorith committed:
Merge remote-tracking branch 'origin/develop' into ui-update

Clorith committed:
Fix merge issue with troubleshootign test

Clorith committed:
Fix js- and stylelints

Clorith committed:
Fix phpunit tests not matching new function names

Clorith committed:
Make grunt tasks slightly more friendly for new contributors, you may just want to get up and running and handle linting and style guides afterwards.
This changes the `default` task to not run linters and such, but only handles dependencies and generating the content of the `build` directory.
Introduces `checkedBuild` as a task that also handles the linting.

Clorith committed:
Fix codestyle violations (#270)

Clorith committed:
Update readme (#271)

Fixes #258

afercia committed:
Improve Show passed tests button. (#265)

afercia committed:
Improve show all themes and plugins buttons accessibility. (#272)

* Improve show all themes and plugins buttons accessibility.
* Add role=list to the list to make Safari happy.

Clorith committed:
Fix heading hierarchy (#277)

Clorith committed:
Add screen reader text to status spans (#278)

Fixes #367, fixes #275

Clorith committed:
Ensure the adminbar menu in troubleshooting mode is keyboard accessible. (#279)

Clorith committed:
Improve contrasts in the troubleshooting dashboard widget (#280)

Clorith committed:
Remove the backup warning modal

Clorith committed:
Fix enqueue conditional missed in previous commit as it only failed outside our plugin page. (#284)

jocastaneda committed:
Merge with latest from develop branch
